File: backend/analysis/hand_value.py
"""牌型价值分析器核心(HandAnalyzer) —— 被 tools/hand_value.py(CLI)、
backend/ai/bot_hv.py(学者Bot)、backend/main.py(分析面板)共用。
模型与口径说明见 tools/hand_value.py 的文件头注释。"""

import math

import numpy as np

from ..native import native

RED = 27

_USEFUL_SET_CACHE = {}


def useful_set(hand) -> list[int]:
    """摸到能降向听(含直接胡)的牌集合。只与手牌有关, 可全局缓存。"""
    key = bytes(hand)
    out = _USEFUL_SET_CACHE.get(key)
    if out is not None:
        return out
    base = native.shanten(list(hand))
    out = []
    for t in range(28):
        if hand[t] >= 4:
            continue
        h = list(hand)
        h[t] += 1
        if native.shanten(h) < base:
            out.append(t)
    _USEFUL_SET_CACHE[key] = out
    return out


class HandAnalyzer:
    def __init__(self, hand, visible_counts, rho=1.0,
                 kaizen=True, kai_margin=2, kai_max=1, kai_topk=6,
                 memo=None, u_eff=None, held_exp=None):
        """hand: 28 计数(当前手牌, 3k+1 或 3k+2 张)。
        visible_counts: 28 计数, 所有可见牌(自己手牌+所有人弃牌+所有副露)。
        rho: 对手摸到你要碰的牌后实际打出来的概率。1.0 = 一定打, 0.0 = 纯自摸。
        kaizen: 换型层 —— 摸到不降向听但让有效张变宽 >= kai_margin 的牌也算进展,
        每条路径最多 kai_max 次(全局预算, 不因降向听而重置 —— 若按"连续次数"
        重置, 每个状态都能花一次预算, 高向听手牌的 DP 图会爆炸, 实测 139s/手)。
        kai_topk: 每个状态最多保留的换型分支数(按进张净增排序) —— 换型层
        会让 DP 状态数从 ~16 爆炸到 ~64 万(单手 0.04s -> 280s), 必须截断;
        0 或 None 表示不截断(仅调试用)。
        memo: 可选外部 dict, 作为 E 的跨实例共享缓存(key 含全部参数)。
        u_eff: 可选 28 向量(允许小数) —— 牌墙中各类牌的期望剩余张数,
        替代默认的均匀假设 u0 = 4 - visible。来自对手牌型分析器:
        u_eff[t] = 不可见[t] - 对手持有期望[t]。只影响自摸/换型通道。
        held_exp: 可选 28 向量 —— 三家对手合计持有各类牌的期望张数。
        给了它, 碰通道权重从 u[t]*3ρ(均匀假设)改为 held_exp[t]*3ρ(按牌细化)。
        两者都不给时行为与旧版完全一致。"""
        self.hand0 = tuple(hand)
        self.u0 = tuple(max(0, 4 - v) for v in visible_counts)
        self.u_eff = (tuple(float(x) for x in u_eff) if u_eff is not None
                      else tuple(float(x) for x in self.u0))
        self.held_exp = tuple(float(x) for x in held_exp) \
            if held_exp is not None else None
        self.rho = rho
        self.kaizen = kaizen
        self.kai_margin = kai_margin
        self.kai_max = kai_max
        self.kai_topk = kai_topk or 99
        self.memo = memo if memo is not None else {}

    # ---------- 有效张(自摸通道) ----------
    def _useful(self, hand, u):
        """摸到能降向听(或胡)的牌 -> {t: 剩余张数}。"""
        s = native.shanten(list(hand))
        return s, {t: u[t] for t in useful_set(hand) if u[t] > 0}

    def _ukeire(self, hand, u):
        return sum(u[t] for t in useful_set(hand) if u[t] > 0)

    # ---------- 碰通道 ----------
    def _peng_transitions(self, hand, u):
        """可碰且有价值的对子 -> {t: (权重, 碰后最优弃牌, 碰后向听)}。
        碰 = 暗手 h[t]-2 成副露, 再从缩短的手牌打出最优一张。
        只在碰后最优向听严格低于当前时才有价值(v31 判碰同口径)。"""
        if self.rho <= 0:
            return {}
        s = native.shanten(list(hand))
        out = {}
        for t in range(27):          # 红中不能被碰
            if hand[t] != 2:
                continue
            if self.held_exp is not None:
                if self.held_exp[t] <= 0:
                    continue
                wt = self.held_exp[t] * 3.0 * self.rho
            else:
                if u[t] <= 0:
                    continue
                wt = u[t] * 3.0 * self.rho
            h2 = list(hand)
            h2[t] -= 2
            # 碰后须打一张: 找最优弃牌(最小向听, 同向听取进张最宽)
            best_d, best_s = -1, 99
            for d in range(28):
                if h2[d] <= 0:
                    continue
                h2[d] -= 1
                sd = native.shanten(h2)
                h2[d] += 1
                if sd < best_s:
                    best_s, best_d = sd, d
            if best_s < s:
                out[t] = (wt, best_d, best_s)
        return out

    def _fast_discard(self, h14, u):
        """摸到有效张后打出哪张: v10 牌效(只保留向听+进张, 关掉两步推演)。"""
        zeros = [0] * 28
        return native.choose_discard_v10(h14, u, zeros, 0.0,
                                         100.0, 1.0, 0.0, 0.0, -1)

    def _kai_candidates(self, hand, u, useful, s):
        """换型候选: [(t, w, gain, h2)], 已按 (-gain,-w) 排序并截断 topk。
        E 与 decompose 共用, 保证口径一致。"""
        kai_tiles = []
        if self.kaizen:
            uk0 = self._ukeire(hand, u)
            u0s = set(t for t, _ in useful)
            cands = []
            for t in range(28):
                if u[t] <= 0 or t in u0s:
                    continue
                h14 = list(hand)
                h14[t] += 1
                d = self._fast_discard(h14, u)
                h14[d] -= 1
                if native.shanten(h14) != s:
                    continue
                gain = self._ukeire(h14, u) - uk0
                if gain >= self.kai_margin:
                    cands.append((gain, t, u[t], tuple(h14)))
            # 只保留进张净增最多的 kai_topk 个分支(状态爆炸的保险丝)
            cands.sort(key=lambda x: (-x[0], -x[2]))
            kai_tiles = cands[: self.kai_topk]
        return kai_tiles

    # ---------- 期望巡数(递推; 巡=轮到自己行动的周期, 碰不消耗摸牌但占一巡) ----------
    def E(self, hand, u, kai=0) -> float:
        key = (hand, u, kai, self.rho, self.kaizen,
               self.kai_margin, self.kai_max, self.kai_topk,
               self.held_exp)
        v = self.memo.get(key)
        if v is not None:
            return v
        s = native.shanten(list(hand))
        useful = [(t, u[t]) for t in useful_set(hand) if u[t] > 0]
        pengs = self._peng_transitions(hand, u)
        kai_tiles = []
        if self.kaizen and kai < self.kai_max:
            kai_tiles = [(t, w, h2) for _g, t, w, h2
                         in self._kai_candidates(hand, u, useful, s)]
        N = sum(u)
        U = (sum(w for _, w in useful) + sum(w for _, w, _ in kai_tiles)
             + sum(w for w, _, _ in pengs.values()))
        if U <= 0:
            val = float(N) + 2.0 * s      # 有效张耗尽的死手
            self.memo[key] = val
            return val
        wait = (N + 1.0) / (U + 1.0)      # 无放回首次命中的精确期望
        val = wait
        for t, w in useful:                # 自摸通道(降向听/胡)
            p = w / U
            h14 = list(hand)
            h14[t] += 1
            if native.is_win(h14):
                continue                  # 这一摸直接胡, 无后续
            u2 = list(u)
            u2[t] -= 1
            d = self._fast_discard(h14, u2)
            h14[d] -= 1
            # kai 透传: 换型预算按整条路径计(不重置), 否则每个状态都能
            # 花一次预算, 高向听手牌的 DP 图会爆炸(实测 139s/手)
            val += p * self.E(tuple(h14), tuple(u2), kai)
        for t, w, h2 in kai_tiles:         # 换型通道: 不降向听但进张变宽
            p = w / U
            u2 = list(u)
            u2[t] -= 1
            val += p * self.E(h2, tuple(u2), kai + 1)
        for t, (w, d, _bs) in pengs.items():   # 碰通道: 对子成副露再弃牌
            p = w / U
            h2 = list(hand)
            h2[t] -= 2
            u2 = list(u)
            u2[t] -= 1
            h2[d] -= 1
            val += p * self.E(tuple(h2), tuple(u2), kai)
        self.memo[key] = val
        return val

    # ---------- E 的通道分解(可解释输出, 与 C mj_hv_explain 同口径) ----------
    def decompose(self, hand) -> dict:
        """hand: 3k+1 计数(出牌后状态)。返回:
        E/wait/c_useful/c_kai/c_peng 五个分量(E = 四项之和, 与 E() 同序累加),
        useful=[(t,剩余)], kai=[(t,剩余,净增)], peng=[(t,权重)] 明细。"""
        hand = tuple(hand)
        u = self.u_eff
        s = native.shanten(list(hand))
        useful = [(t, u[t]) for t in useful_set(hand) if u[t] > 0]
        pengs = self._peng_transitions(hand, u)
        kai_tiles = self._kai_candidates(hand, u, useful, s) \
            if self.kai_max > 0 else []
        N = sum(u)
        U = (sum(w for _, w in useful)
             + sum(w for _g, _t, w, _h in kai_tiles)
             + sum(w for w, _, _ in pengs.values()))
        cU = cK = cP = 0.0
        if U <= 0:
            wait = float(N) + 2.0 * s
            total = wait
        else:
            wait = (N + 1.0) / (U + 1.0)
            total = wait
            for t, w in useful:
                p = w / U
                h14 = list(hand)
                h14[t] += 1
                if native.is_win(h14):
                    continue
                u2 = list(u)
                u2[t] -= 1
                d = self._fast_discard(h14, u2)
                h14[d] -= 1
                c = p * self.E(tuple(h14), tuple(u2), 0)
                total += c
                cU += c
            for _g, t, w, h2 in kai_tiles:
                p = w / U
                u2 = list(u)
                u2[t] -= 1
                c = p * self.E(h2, tuple(u2), 1)
                total += c
                cK += c
            for t, (w, d, _bs) in pengs.items():
                p = w / U
                h2 = list(hand)
                h2[t] -= 2
                u2 = list(u)
                u2[t] -= 1
                h2[d] -= 1
                c = p * self.E(tuple(h2), tuple(u2), 0)
                total += c
                cP += c
        return {"E": total, "wait": wait, "c_useful": cU, "c_kai": cK,
                "c_peng": cP, "useful": useful,
                "kai": [(t, w, g) for g, t, w, _h in kai_tiles],
                "peng": [(t, w) for t, (w, _d, _s) in pengs.items()],
                "shanten": s}

    # ---------- n 进张内可达和牌型 ----------
    def enum_patterns(self, start_hand, max_draws):
        """DFS: 只走有效摸牌 + 保持最小向听的弃牌。按所需摸牌组合聚合。"""
        h0 = list(start_hand)
        assert sum(h0) % 3 == 1, "枚举从 3k+1 张(出牌后)状态开始"
        agg = {}     # needed_tuple -> {"patterns": set, "tiles": [(t,c)]}
        nodes = [0]

        def rec(hand, u, depth):
            nodes[0] += 1
            if nodes[0] > 300000 or depth > max_draws:
                return
            s, useful = self._useful(hand, u)
            for t in useful:
                h14 = list(hand)
                h14[t] += 1
                if native.is_win(h14):
                    needed = tuple(max(0, h14[i] - h0[i]) for i in range(28))
                    ent = agg.setdefault(needed, {"n_pat": 0})
                    ent["n_pat"] += 1
                    continue
                if depth == max_draws:
                    continue
                # 保持最小向听的弃牌才值得继续
                opts = []
                best = 99
                for d in range(28):
                    if h14[d] <= 0:
                        continue
                    h14[d] -= 1
                    sd = native.shanten(h14)
                    h14[d] += 1
                    if sd < best:
                        best = sd
                        opts = [d]
                    elif sd == best:
                        opts.append(d)
                u2 = list(u)
                u2[t] -= 1
                for d in opts:
                    h2 = list(h14)
                    h2[d] -= 1
                    rec(tuple(h2), tuple(u2), depth + 1)

        rec(tuple(h0), tuple(self.u0), 1)
        out = []
        for needed, ent in agg.items():
            tiles = [(t, c) for t, c in enumerate(needed) if c > 0]
            out.append({"tiles": tiles, "n_pat": ent["n_pat"],
                        "draws": sum(c for _, c in tiles)})
        out.sort(key=lambda x: (x["draws"], -x["n_pat"]))
        return out, nodes[0]

    # ---------- 单组合独立期望(包含-排除) ----------
    def pattern_time(self, tiles):
        """集齐所需组合 {(t,c)} 的期望摸牌数(独立地看这一个组合)。
        对子成刻(hand[t]==2 且需要第3张)的牌可用量 ×(1+3ρ)(若开启补偿)。"""
        if len(tiles) > 4 or any(c > 2 for _, c in tiles):
            return None
        h0 = self.hand0
        mult = 1.0 + 3.0 * self.rho
        a = []
        for t, c in tiles:
            av = self.u0[t]
            if self.rho > 0 and h0[t] == 2 and c == 1 and t != RED:
                av = min(int(round(av * mult)), 16)   # 速率×(1+3ρ) ≈ 可用量放大(近似)
            a.append(av)
        N = sum(self.u0)
        m = len(tiles)
        E = 0.0
        # E[T] = sum_k P(T > k); P(T>k) = P(存在某类牌没集齐) = 包含-排除
        for k in range(0, N + 1):
            pin = 0.0
            for mask in range(1, 1 << m):
                bits = bin(mask).count("1")
                sign = 1.0 if bits % 2 == 1 else -1.0
                A = sum(a[i] for i in range(m) if mask >> i & 1)
                # 子集 S 全部未集齐: 每类抽到的数 x_i < c_i
                ranges = [range(0, tiles[i][1]) for i in range(m)
                          if mask >> i & 1]
                idx = [i for i in range(m) if mask >> i & 1]
                tot = 0.0

                def rec2(j, xsum, prod):
                    if j == len(idx):
                        if k - xsum < 0 or k - xsum > N - A:
                            return
                        nonlocal_tot[0] += prod * \
                            math.comb(N - A, k - xsum)
                        return
                    i = idx[j]
                    for x in ranges[j]:
                        rec2(j + 1, xsum + x, prod * math.comb(a[i], x))

                nonlocal_tot = [0.0]
                rec2(0, 0, 1.0)
                pin += sign * nonlocal_tot[0] / math.comb(N, k)
            E += pin
        return E

    # ---------- Monte Carlo 对照(单人摸打模拟) ----------
    def mc(self, start_hand, n_sims=5000, seed=1):
        pool0 = [t for t in range(28) for _ in range(self.u0[t])]
        rng = np.random.default_rng(seed)
        counts = []
        h0 = list(start_hand)
        assert sum(h0) % 3 == 1
        for _ in range(n_sims):
            pool = rng.permutation(pool0)
            h = list(h0)
            u = list(self.u0)
            done = False
            for i, t in enumerate(pool):
                h[t] += 1
                u[t] -= 1
                if native.is_win(h):
                    counts.append(i + 1)
                    done = True
                    break
                d = self._fast_discard(h, u)
                h[d] -= 1
            if not done:
                counts.append(len(pool) + 2)
        return float(np.mean(counts))

File: android/app/src/main/assets/js/hand_value.js
/* 安康159 - 牌型价值分析器(HandAnalyzer, JS版)
 *
 * 与 backend/analysis/hand_value.py 逐行对齐, 任何一侧修改必须同步另一侧,
 * 并用跨语言单测验证 0 不一致。
 *
 * 模型: E(hand, u) = 期望巡数(巡 = 轮到自己行动的周期; 碰不消耗摸牌但占一巡)。
 *   递推 = 首次命中有效张的等待巡数 + 各通道命中后的后续期望之和。
 *   三条通道:
 *     自摸通道  摸到能降向听(或直接胡)的牌
 *     换型通道  摸到不降向听但让有效张变宽 >= kaiMargin 的牌(kaizen 层)
 *     碰通道    对子被打出后碰成副露, 再打一张
 *
 * 只移植了学者Bot 需要的部分。Python 侧的 enum_patterns / pattern_time / mc
 * 是 CLI 交互分析用的(依赖 numpy 与包含-排除组合数), 移动端不需要, 未移植。
 *
 * 性能取舍(与 Python 一致): 换型层(kaizen)把单手分析从毫秒级拉到秒级,
 * 只改善绝对值不改善排序, 所以 Bot 对战按 kaizen=false 跑。
 *
 * 【有效张口径】Python 侧其实存在两种 ukeire 定义:
 *   hand_value.useful_set  一律用 shanten(h+t) < base
 *   C 的 tiles_info      对 s==0 特判成 is_win(h+t), 否则 shanten 降低
 * _fast_discard 走 C 口径, _ukeire 走 useful_set 口径。两者在 3k+1 与 3k+2
 * 形态上实测等价(听牌时加牌能胡 <=> 向听降到 -1), 仅在 3k+0 形态上不等价。
 * 而 E() 只会从 3k+1 手牌出发(h14 = hand+1 为 3k+2), 永远不会碰到 3k+0,
 * 所以本文件统一用 usefulSet 一种定义就够了。
 * 注: 若向 E() 传入 3k+2 手牌(越调用约定), h14 会变成 3k+0, 两端会出现
 * 分歧 —— 写测试时必须用对的手牌张数(弃牌用 14-3n, 应对用 13-3n)。
 *
 * 依赖: tiles.js(TILE_COUNT/RED), win.js(shanten/isWin)
 */

/* ================= 有效张集合(全局缓存) ================= */

const _usefulSetCache = new Map();
const USEFUL_SET_CACHE_LIMIT = 200000;

// 注: 曾试过再加两层缓存(顶层 shanten 结果缓存 + _fastDiscard 结果缓存),
// 实测毫无改善(4582ms vs 4781ms, 在噪声内) —— E 递归访问的状态大多互不重复,
// 瓶颈是状态总量而非重复查询。已移除, 避免无收益的复杂度。

/** 28 计数的缓存键。值域 0..4, 用 charCode 拼串比 join(",") 快很多 */
function _hvKey(c) {
  return String.fromCharCode(
    c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7], c[8], c[9],
    c[10], c[11], c[12], c[13], c[14], c[15], c[16], c[17], c[18], c[19],
    c[20], c[21], c[22], c[23], c[24], c[25], c[26], c[27]);
}

/**
 * 摸到能降向听(含直接胡)的牌集合, 升序。只与手牌有关, 可全局缓存。
 * 对齐 hand_value.py 的 useful_set。
 * @param {number[]} hand 长度 28 的计数
 * @returns {number[]} 升序 tile 列表(返回值被缓存共享, 调用方不得修改)
 */
function usefulSet(hand) {
  const key = _hvKey(hand);
  const hit = _usefulSetCache.get(key);
  if (hit !== undefined) return hit;
  const base = shanten(hand);
  const out = [];
  for (let t = 0; t < 28; t++) {
    if (hand[t] >= 4) continue;
    hand[t]++;
    const s = shanten(hand);
    hand[t]--;
    if (s < base) out.push(t);
  }
  // 达到上限整体 clear(), 不能"满了就不写" —— 那会永久退化为零缓存
  if (_usefulSetCache.size >= USEFUL_SET_CACHE_LIMIT) _usefulSetCache.clear();
  _usefulSetCache.set(key, out);
  return out;
}

/* ================= 分析器 ================= */

class HandAnalyzer {
  /**
   * @param {number[]} hand 28 计数(当前手牌, 3k+1 或 3k+2 张)
   * @param {number[]} visibleCounts 28 计数, 所有可见牌(自己手牌+所有人弃牌+所有副露)
   * @param {object} [opts]
   *   rho     对手摸到你要碰的牌后实际打出来的概率。1=一定打, 0=纯自摸
   *   kaizen  换型层开关(Bot 对战用 false)
   *   kaiMargin / kaiMax  换型判定阈值与每条路径的最大换型次数(全局预算,
   *           不随降向听重置; 若按连续次数重置会让高向听手牌 DP 图爆炸)
   *   kaiTopk 每状态最多保留的换型分支数(按进张净增排序)。换型层会让
   *           DP 状态数从 ~16 爆炸到 ~64 万(单手 0.04s -> 280s), 必须截断;
   *           传 0 表示不截断(仅调试用)
   */
  constructor(hand, visibleCounts, opts) {
    const o = opts || {};
    this.hand0 = hand.slice();
    this.u0 = new Array(28);
    for (let i = 0; i < 28; i++) {
      const v = 4 - visibleCounts[i];
      this.u0[i] = v > 0 ? v : 0;
    }
    this.rho = o.rho !== undefined ? o.rho : 1.0;
    this.kaizen = o.kaizen !== undefined ? o.kaizen : true;
    this.kaiMargin = o.kaiMargin !== undefined ? o.kaiMargin : 2;
    this.kaiMax = o.kaiMax !== undefined ? o.kaiMax : 1;
    this.kaiTopk = o.kaiTopk !== undefined ? o.kaiTopk : 6;
    this.memo = new Map();
  }

  /** 有效张总张数(任意向听下) */
  _ukeire(hand, u) {
    let sum = 0;
    for (const t of usefulSet(hand)) if (u[t] > 0) sum += u[t];
    return sum;
  }

  /**
   * 可碰且有价值的对子 -> [[t, 权重, 碰后最优弃牌, 碰后向听], ...] (t 升序)
   *
   * 碰 = 暗手 h[t]-2 成副露, 再从缩短的手牌打出最优一张。
   * 只在碰后最优向听严格低于当前时才有价值(与 v31 判碰同口径)。
   *
   * 顺序必须保持 t 升序 —— E() 里按此顺序累加浮点, 换序会产生末位差异。
   */
  _pengTransitions(hand, u) {
    if (this.rho <= 0) return [];
    const s = shanten(hand);
    const out = [];
    for (let t = 0; t < 27; t++) {          // 红中不能被碰
      if (hand[t] !== 2 || u[t] <= 0) continue;
      const h2 = hand.slice();
      h2[t] -= 2;
      // 碰后须打一张: 找最优弃牌(最小向听)
      let bestD = -1, bestS = 99;
      for (let d = 0; d < 28; d++) {
        if (h2[d] <= 0) continue;
        h2[d]--;
        const sd = shanten(h2);
        h2[d]++;
        if (sd < bestS) { bestS = sd; bestD = d; }
      }
      if (bestS < s) out.push([t, u[t] * 3.0 * this.rho, bestD, bestS]);
    }
    return out;
  }

  /**
   * 摸到有效张后打出哪张: v10 牌效(只保留向听+进张, 关掉两步推演与放杠风险)。
   *
   * 等价于 Python 的
   *   native.choose_discard_v10(h14, u, zeros, 0.0, 100.0, 1.0, 0.0, 0.0, -1)
   * 即 C 里 score_discards_v10 在 SW=100 UW=1 CW=0 RW=0 cont_max=-1 下的行为:
   *   - 向听非最小的候选: score = -10*SW - SW*s (恒为负, 必然落选)
   *   - 向听最小的候选:   score = UW * ukeire
   *   - 取最大, 严格 > 比较且按 tile 升序扫描 -> 同分取最小 tile
   * (cont_max=-1 时 s<=cont_max 不成立, 因为 3k+1 手牌向听 >= 0, 故 cont 恒为 0)
   */
  _fastDiscard(h14, u) {
    const cand = [], cs = [];
    let minSh = 99;
    for (let t = 0; t < 28; t++) {
      if (h14[t] <= 0) continue;
      h14[t]--;
      const s = shanten(h14);
      h14[t]++;
      cand.push(t); cs.push(s);
      if (s < minSh) minSh = s;
    }
    if (!cand.length) return -1;
    let bestT = -1, bestScore = -1e18;
    for (let k = 0; k < cand.length; k++) {
      const t = cand[k], s = cs[k];
      let score;
      if (s > minSh) {
        score = -10.0 * 100.0 - 100.0 * s;
      } else {
        h14[t]--;
        score = 1.0 * this._ukeire(h14, u);
        h14[t]++;
      }
      if (score > bestScore) { bestScore = score; bestT = t; }
    }
    return bestT;
  }

  /**
   * 期望巡数。hand/u 均为长度 28 的计数数组, 不会被修改。
   *
   * 浮点累加顺序必须与 Python 完全一致(wait -> 自摸通道 -> 换型通道 -> 碰通道,
   * 每条通道内按 t 升序), 否则末位会出现差异。
   */
  E(hand, u, kai) {
    if (kai === undefined) kai = 0;
    const key = _hvKey(hand) + "|" + _hvKey(u) + "|" + kai;
    const hit = this.memo.get(key);
    if (hit !== undefined) return hit;

    const s = shanten(hand);
    const useful = [];
    for (const t of usefulSet(hand)) if (u[t] > 0) useful.push([t, u[t]]);
    const pengs = this._pengTransitions(hand, u);

    const kaiTiles = [];
    if (this.kaizen && kai < this.kaiMax) {
      const uk0 = this._ukeire(hand, u);
      const inUseful = new Set();
      for (const [t] of useful) inUseful.add(t);
      const cands = [];
      for (let t = 0; t < 28; t++) {
        if (u[t] <= 0 || inUseful.has(t)) continue;
        const h14 = hand.slice();
        h14[t]++;
        const d = this._fastDiscard(h14, u);
        h14[d]--;
        if (shanten(h14) !== s) continue;
        const gain = this._ukeire(h14, u) - uk0;
        if (gain >= this.kaiMargin) {
          cands.push([gain, t, u[t], h14]);
        }
      }
      // 只保留进张净增最多的 kaiTopk 个分支(状态爆炸的保险丝)
      // 排序与 Python 的 (-gain, -w) 稳定排序(t 升序插入)完全一致
      cands.sort((a, b) => (b[0] - a[0]) || (b[2] - a[2]) || (a[1] - b[1]));
      const cap = this.kaiTopk > 0 ? this.kaiTopk : cands.length;
      for (const [, t, w, h2] of cands.slice(0, cap)) kaiTiles.push([t, w, h2]);
    }

    let N = 0;
    for (let i = 0; i < 28; i++) N += u[i];
    // 与 Python 的 (sum(useful) + sum(kai)) + sum(peng) 同序: 前两项整数, 末项浮点
    let sU = 0;
    for (const [, w] of useful) sU += w;
    let sK = 0;
    for (const kt of kaiTiles) sK += kt[1];
    let sP = 0;
    for (const pg of pengs) sP += pg[1];
    const U = sU + sK + sP;

    if (U <= 0) {
      const val = N + 2.0 * s;            // 有效张耗尽的死手
      this.memo.set(key, val);
      return val;
    }

    let val = (N + 1.0) / (U + 1.0);      // 无放回首次命中的精确期望

    for (const [t, w] of useful) {        // 自摸通道(降向听/胡)
      const p = w / U;
      const h14 = hand.slice();
      h14[t]++;
      if (isWin(h14)) continue;           // 这一摸直接胡, 无后续
      const u2 = u.slice();
      u2[t]--;
      const d = this._fastDiscard(h14, u2);
      h14[d]--;
      // kai 透传: 换型预算按整条路径计(不重置), 否则每个状态都能
      // 花一次预算, 高向听手牌的 DP 图会爆炸(实测 139s/手)
      val += p * this.E(h14, u2, kai);
    }

    for (const [t, w, h2] of kaiTiles) {  // 换型通道: 不降向听但进张变宽
      const p = w / U;
      const u2 = u.slice();
      u2[t]--;
      val += p * this.E(h2, u2, kai + 1);
    }

    for (const [t, w, d] of pengs) {      // 碰通道: 对子成副露再弃牌
      const p = w / U;
      const h2 = hand.slice();
      h2[t] -= 2;
      const u2 = u.slice();
      u2[t]--;
      h2[d]--;
      val += p * this.E(h2, u2, kai);
    }

    this.memo.set(key, val);
    return val;
  }
}
